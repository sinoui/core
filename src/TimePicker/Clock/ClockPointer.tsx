import React from 'react';
import styled from 'styled-components';
import { CLOCK_NUMBER_PADDING } from './constants';

const Wrapper = styled.div<{
  $size: number;
}>`
  position: absolute;
  left: calc((100% - 2px) / 2);
  top: ${CLOCK_NUMBER_PADDING}px;
  width: 2px;
  height: ${({ $size }) => $size / 2 - CLOCK_NUMBER_PADDING}px;
  background-color: ${({ theme }) => theme.palette.primary.main};

  transform-origin: bottom;
  .sinoui-clock__pointer {
    position: absolute;
    left: calc((100% - 32px) / 2);
    top: 0;
    width: 32px;
    height: 32px;
    user-select: none;
    border-radius: 50%;
    background-color: ${({ theme }) => theme.palette.primary.main};
  }
`;

export default function ClockPointer({ size }: { size: number }) {
  return (
    <Wrapper $size={size}>
      <div className="sinoui-clock__pointer" />
    </Wrapper>
  );
}
