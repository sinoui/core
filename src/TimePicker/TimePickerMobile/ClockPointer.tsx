import React, { useRef, useEffect } from 'react';
import styled from 'styled-components';

const Wrapper = styled.div<{
  $rotateDeg: number;
  $isTransition: boolean;
}>`
  position: absolute;
  left: calc((100% - 2px) / 2);
  width: 2px;
  height: 40%;
  bottom: 50%;
  background-color: ${({ theme }) => theme.palette.primary.main};
  ${(props) =>
    props.$isTransition &&
    `transition: ${props.theme.transitions.create('transform')}`};
  transform: ${({ $rotateDeg = 0 }) => `rotate3d(0,0,1,${$rotateDeg}deg)`};
  transform-origin: center bottom 0;
`;

const MinuteDot = styled.div<{
  $selected?: boolean;
}>`
  top: -21px;
  left: -15px;
  width: 4px;
  border: 14px solid ${({ theme }) => theme.palette.primary.main};
  height: 4px;
  position: absolute;
  box-sizing: content-box;
  border-radius: 100%;
  background-color: ${({ theme, $selected }) =>
    $selected ? '#fff' : theme.palette.primary.main};
`;

export default function ClockPointer({
  rotateDeg,
  isHourView,
  isTransition,
  onTransitionChange,
  ...rest
}: {
  rotateDeg: number;
  isHourView: boolean;
  isTransition: boolean;
  onTransitionChange: (transitionValue: boolean) => void;
}) {
  const hourViewRef = useRef(isHourView);

  useEffect(() => {
    // 时分视图切换时 增加指针旋转动画
    if (hourViewRef.current !== isHourView) {
      onTransitionChange(true);
    }
    hourViewRef.current = isHourView;
  }, [isHourView, onTransitionChange]);

  return (
    <Wrapper $rotateDeg={rotateDeg} $isTransition={isTransition} {...rest}>
      <MinuteDot
        className="sinoui-clock__pointer-minute-dot"
        $selected={!isHourView && rotateDeg % 5 !== 0}
      />
    </Wrapper>
  );
}
