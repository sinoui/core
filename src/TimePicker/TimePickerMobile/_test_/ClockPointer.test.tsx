import React from 'react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import ClockPointer from '../ClockPointer';

it('分钟视图，旋转到11分，圆点背景色为白色', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <ClockPointer
        isHourView={false}
        rotateDeg={36}
        data-testid="clock-pointer"
      />
    </ThemeProvider>,
  );
  expect(
    getByTestId('clock-pointer').querySelector(
      '.sinoui-clock__pointer-minute-dot',
    ),
  ).toHaveStyle('background-color: #fff');
});

it('分钟视图，旋转到10分，圆点背景色为主题色', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <ClockPointer
        isHourView={false}
        rotateDeg={30}
        data-testid="clock-pointer"
      />
    </ThemeProvider>,
  );
  expect(
    getByTestId('clock-pointer').querySelector(
      '.sinoui-clock__pointer-minute-dot',
    ),
  ).toHaveStyle('background-color: #3f51b5');
});

it('时钟视图，圆点背景色为主题色', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <ClockPointer isHourView rotateDeg={30} data-testid="clock-pointer" />
    </ThemeProvider>,
  );
  expect(
    getByTestId('clock-pointer').querySelector(
      '.sinoui-clock__pointer-minute-dot',
    ),
  ).toHaveStyle('background-color: #3f51b5');
});

describe('快照测试', () => {
  it('时钟视图', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <ClockPointer isHourView rotateDeg={90} />
        </ThemeProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('分钟视图', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <ClockPointer isHourView={false} rotateDeg={180} />
        </ThemeProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
