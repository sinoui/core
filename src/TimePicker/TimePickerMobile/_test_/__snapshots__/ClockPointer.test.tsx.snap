// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 分钟视图 1`] = `
<div
  className="sc-bdnylx kzDLEo"
>
  <div
    className="sc-gtssRu hPKbEG sinoui-clock__pointer-minute-dot"
  />
</div>
`;

exports[`快照测试 时钟视图 1`] = `
<div
  className="sc-bdnylx eAeEDI"
>
  <div
    className="sc-gtssRu hPKbEG sinoui-clock__pointer-minute-dot"
  />
</div>
`;
